pkg/kubelet/kuberuntime/kuberuntime_logs.go

```diff
@@ -0,0 +1,332 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberuntime
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/docker/docker/pkg/jsonlog"
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// Notice that the current kuberuntime logs implementation doesn't handle
+// log rotation.
+// * It will not retrieve logs in rotated log file.
+// * If log rotation happens when following the log:
+//   * If the rotation is using create mode, we'll still follow the old file.
+//   * If the rotation is using copytruncate, we'll be reading at the original position and get nothing.
+// TODO(random-liu): Support log rotation.
+
+// streamType is the type of the stream.
+type streamType string
+
+const (
+	stderrType streamType = "stderr"
+	stdoutType streamType = "stdout"
+
+	// timeFormat is the time format used in the log.
+	timeFormat = time.RFC3339Nano
+	// blockSize is the block size used in tail.
+	blockSize = 1024
+)
+
+var (
+	// eol is the end-of-line sign in the log.
+	eol = []byte{'\n'}
+	// delimiter is the delimiter for timestamp and streamtype in log line.
+	delimiter = []byte{' '}
+)
+
+// logMessage is the internal log type.
+type logMessage struct {
+	timestamp time.Time
+	stream    streamType
+	log       []byte
+}
+
+// reset resets the log to nil.
+func (l *logMessage) reset() {
+	l.timestamp = time.Time{}
+	l.stream = ""
+	l.log = nil
+}
+
+// logOptions is the internal type of all log options.
+type logOptions struct {
+	tail      int64
+	since     time.Time
+	follow    bool
+	timestamp bool
+}
+
+// newLogOptions convert the api.PodLogOptions to internal logOptions.
+func newLogOptions(apiOpts *api.PodLogOptions, now time.Time) *logOptions {
+	opts := &logOptions{
+		tail:      -1, // -1 by default which means read all logs.
+		follow:    apiOpts.Follow,
+		timestamp: apiOpts.Timestamps,
+	}
+	if apiOpts.TailLines != nil {
+		opts.tail = *apiOpts.TailLines
+	}
+	if apiOpts.SinceSeconds != nil {
+		opts.since = now.Add(-time.Duration(*apiOpts.SinceSeconds) * time.Second)
+	}
+	if apiOpts.SinceTime != nil && apiOpts.SinceTime.After(opts.since) {
```

Comment:
maxBytes is particularly important for UIs to avoid blowing out browsers - doesn't have to be 100% accurate (original design allowed maxbytes to be slightly inaccurate), but is important to prevent DoS on kubelet as well.